rust/src/bwrap.rs

```diff
@@ -0,0 +1,481 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use crate::cxxrsutil::*;
+use crate::ffi::BubblewrapMutability;
+use anyhow::{Context, Result};
+use fn_error_context::context;
+use openat_ext::OpenatDirExt;
+use std::convert::TryInto;
+use std::num::NonZeroUsize;
+use std::os::unix::io::AsRawFd;
+use std::path::Path;
+use std::pin::Pin;
+use std::process::Command;
+
+// Links in the rootfs to /usr
+static USR_LINKS: [&str; 5] = ["lib", "lib32", "lib64", "bin", "sbin"];
+// This is similar to what systemd does, except we drop /usr/local, since scripts shouldn't see it.
+static PATH_VAR: &str = "PATH=/usr/sbin:/usr/bin";
+/// Explicitly added capabilities.  See the bits that call --cap-drop below for more
+/// information.
+static ADDED_CAPABILITIES: &[&str] = &[
+    "cap_chown",
+    "cap_dac_override",
+    "cap_fowner",
+    "cap_fsetid",
+    "cap_kill",
+    "cap_setgid",
+    "cap_setuid",
+    "cap_setpcap",
+    "cap_sys_chroot",
+    "cap_setfcap",
+];
+
+/// Filesystems explicitly mounted readonly; these were cargo culted from
+/// docker/podman I think at some point.
+static RO_BINDS: &[&str] = &[
+    "/sys/block",
+    "/sys/bus",
+    "/sys/class",
+    "/sys/dev",
+    "/sys/devices",
+];
+
+pub(crate) struct Bubblewrap {
+    pub(crate) rootfs_fd: openat::Dir,
+
+    executed: bool,
+    argv: Vec<String>,
+    child_argv0: Option<NonZeroUsize>,
+    launcher: gio::SubprocessLauncher, // 🚀
+
+    rofiles_mounts: Vec<RoFilesMount>,
+}
+
+// nspawn by default doesn't give us CAP_NET_ADMIN; see
+// https://pagure.io/releng/issue/6602#comment-71214
+// https://pagure.io/koji/pull-request/344#comment-21060
+// Theoretically we should do capable(CAP_NET_ADMIN)
+// but that's a lot of ugly code, and the only known
+// place we hit this right now is nspawn.  Plus
+// we want to use userns down the line anyways where
+// we'll regain CAP_NET_ADMIN.
+fn running_in_nspawn() -> bool {
+    std::env::var_os("container").as_deref() == Some(std::ffi::OsStr::new("systemd-nspawn"))
+}
+
+/// A wrapper for rofiles-fuse from ostree.  This protects the underlying
+/// hardlinked files from mutation.  The mount point is a temporary
+/// directory.
+struct RoFilesMount {
+    /// Our inner tempdir; this is only an Option<T> so we can take ownership
+    /// of it in drop().
+    tempdir: Option<tempfile::TempDir>,
+}
+
+impl RoFilesMount {
+    /// Create a new rofiles-fuse mount point
+    fn new(rootfs: &openat::Dir, path: &str) -> Result<Self> {
+        let path = path.trim_start_matches('/');
+        let tempdir = tempfile::Builder::new()
+            .prefix("rpmostree-rofiles-fuse")
+            .tempdir()?;
+        let status = std::process::Command::new("rofiles-fuse")
+            .arg("--copyup")
+            .arg(path)
+            .arg(tempdir.path())
+            .current_dir(format!("/proc/self/fd/{}", rootfs.as_raw_fd()))
+            .status()?;
+        if !status.success() {
+            return Err(anyhow::anyhow!("{}", status));
```

Comment:
Should we print the full cmdline that failed here? [This](https://github.com/coreos/coreos-installer/blob/2d17136f5c7a6e57835196b42c3bcdad6f051f18/src/util.rs#L18-L34) might be useful to copy.

Reply:
That'd be *really* verbose here.